internal/compiler/compiler.go

```diff
@@ -710,15 +728,12 @@ func (c *compiler) expr(expr ast.Expr) {
 			c.expr(&ast.NumExpr{1})
 			c.add(op)
 			c.add(Dupe)
+			c.assign(e.Expr)
 		} else {
-			c.expr(e.Expr)
+			c.incrExpr(e)
```

Comment:
You've actually uncovered (or made me notice) a larger problem with GoAWK here! All the augmented assignment operations that assign to an lvalue -- `++` / `--`, `+=` and other augmented assignment operators, and `sub` / `gsub` (because they can assign to an lvalue) -- are buggy when used as expressions too.

They evaluate the field/array index twice. For example, in the above, `c.expr(e.Expr)` evaluates the field/array index for FieldExpr/IndexExpr, and the `c.assign(e.Expr)` does it again. This is a bug.

I've added a bunch of tests [in this Gist](https://gist.github.com/benhoyt/437f59a796d4c9bdaa982960112a6c1c) that show this.

So we actually need to step back a bit and fix it in `expr` for `IncrExpr`, `AugAssignExpr`, and `CallExpr` for the `sub` / `gsub` cases. I had a play with this today, but it's not trivial!

My current thinking is to not use `c.expr()` directly on the expression, but use a function with only handles lvalues and saves the index away somewhere to be used by `c.assign` later. But I'm not sure -- there's bound to be a more elegant way to solve this.

If you'd like to suggest a good way to fix all of these, go ahead. Otherwise we can reduce this PR to just fixing the parsing issue and I'll address the compiler bugs in a separate PR.

Reply:
One possible solution could be to introduce a data structure on the virtual machine side where you can save the indices from time to time and then use them within `c.assign` by fetching them with specific instructions.

Another type of solution could be to create functions similar to my `incrExpr` that handle all cases (one for `IncrExpr` with pre-increment, one for `IncrExpr` with post-increment, one for `AugAssignExpr`, and one for `sub` and `gsub`).

A final approach (which is what I committed) is to save the index on the stack and retrieve it using the Roll instruction (similar to the roll instruction in Forth), which allows you to move an element from a lower position on the stack to the top. So, essentially, the solution is similar to the first one, but instead of creating a new data structure, the stack is reused. I understand that this approach might be inefficient as you need to "shift" the stack by a few elements each time you call Roll.

Reply:
I think the code can be merged. I just added one test for nested gsub expressions (I think this case is worth testing, since gsub/sub, as opposed to the other augmented assignment operators, pushes two values on the stack)